website/blog/2024-08-06-langgraph-model-from-code/index.md

```diff
@@ -209,54 +206,81 @@ In the code below, we demonstrate that our chain has chatbot functionality!
-response = loaded_model.invoke(payload)
-
-print(f"User: {message}")
-print(f"Agent: {get_most_recent_message(response)}")
-
-print("\n-------- Message 3 -----------")
-message = "What is my name?"
-new_messages = increment_message_history(response, {"role": "user", "content": message})
-payload = {"messages": new_messages}
-response = loaded_model.invoke(payload)
-
-print(f"User: {message}")
-print(f"Agent: {get_most_recent_message(response)}")
+with mlflow.start_run():
+    loaded_model = mlflow.langchain.load_model(model_uri)
+
+    # Show inference and message history functionality
+    print("-------- Message 1 -----------")
+    message = "What's my name?"
+    payload = {"messages": [{"role": "user", "content": message}]}
+    response = loaded_model.invoke(payload)
+
+    print(f"User: {message}")
+    print(f"Agent: {get_most_recent_message(response)}")
+
+    print("\n-------- Message 2 -----------")
+    message = "My name is Morpheus."
+    new_messages = increment_message_history(response, {"role": "user", "content": message})
+    payload = {"messages": new_messages}
+    response = loaded_model.invoke(payload)
+
+    print(f"User: {message}")
+    print(f"Agent: {get_most_recent_message(response)}")
+
+    print("\n-------- Message 3 -----------")
+    message = "What is my name?"
+    new_messages = increment_message_history(response, {"role": "user", "content": message})
+    payload = {"messages": new_messages}
+    response = loaded_model.invoke(payload)
+
+    print(f"User: {message}")
+    print(f"Agent: {get_most_recent_message(response)}")
 ```
 
 Ouput:
 
 ```text
 -------- Message 1 -----------
 User: What's my name?
-Agent: I'm sorry, I don't know your name. Can you please tell me?
+Agent: I'm sorry, I cannot guess your name as I do not have access to that information. If you would like to share your name with me, feel free to do so.
 
 -------- Message 2 -----------
 User: My name is Morpheus.
 Agent: Nice to meet you, Morpheus! How can I assist you today?
 
 -------- Message 3 -----------
-User: What's my name?
-Agent: Your name is Morpheus!
+User: What is my name?
+Agent: Your name is Morpheus.
+```
+
+### 4.1 - MLflow Tracing
+
+Before concluding, let's demonstrate [MLflow tracing](https://mlflow.org/docs/latest/llms/tracing/index.html).
+
+MLflow Tracing is a feature that enhances LLM observability in your Generative AI (GenAI) applications by capturing detailed information about the execution of your application’s services. Tracing provides a way to record the inputs, outputs, and metadata associated with each intermediate step of a request, enabling you to easily pinpoint the source of bugs and unexpected behaviors.
+
+Let's enter the mlflow UI by running the below command in your working directory. It must contain the auto-generated `mlruns` directory.
+
+```shell
+mlflow ui
 ```
 
+With that running, we can go to the provided url which defaults to `http://127.0.0.1:5000` in our browser. This is the MLflow UI.
```

Comment:
I think we can avoid all of this and just point to the 'tracking server' guide in the docs. This will run a local instance of the UI, which is confusing for users that already have a running tracking server or use a managed service like Databricks, SageMaker, or AzureML.

Reply:
We can just say "Navigate to the experiment that we have set as active within the MLflow UI to see the traces tab" (or something like that)